### PR TITLE
gh-113440: Windows CI: Raise exception with stderr if `ver` call fails

### LINKED ISSUE #113440: Windows 32-bit CI: failures calling `ver` command
We've had build failures on "Windows / build and test (x86)", after some unknown problem calling the  `ver` command.

For example:

```pytb
Run .\python.bat -m test.pythoninfo
Running Debug|Win32 interpreter...
ERROR: collect_windows() failed
Traceback (most recent call last):
  File "D:\a\cpython\cpython\Lib\test\pythoninfo.py", line 1022, in collect_info
    collect_func(info_add)
    ~~~~~~~~~~~~^^^^^^^^^^
  File "D:\a\cpython\cpython\Lib\test\pythoninfo.py", line [9](https://github.com/python/cpython/actions/runs/7286334537/job/19854864684#step:4:10)34, in collect_windows
    line = output.splitlines()[0]
           ~~~~~~~~~~~~~~~~~~~^^^
IndexError: list index out of range

Collection failed: exit with error
```

https://github.com/python/cpython/actions/runs/7286334537/job/19854864684

More examples:

* https://github.com/python/cpython/actions/runs/7298148666/job/19888624148?pr=113378
* https://github.com/python/cpython/actions/runs/7298041643/job/19888309424?pr=113390
* https://github.com/python/cpython/actions/runs/7297567068/job/19887032026?pr=113389

Sometimes it fails on re-run, sometimes it passes.

### Linked PRs
* gh-113392
* gh-113435
* gh-113452
* gh-113453

### PR DESCRIPTION
**Pull request by hugovk**
Fixes gh-113440.

We've had build failures on "Windows / build and test (x86)", after some unknown problem calling the  `ver` command.

For example:

```pytb
Run .\python.bat -m test.pythoninfo
Running Debug|Win32 interpreter...
ERROR: collect_windows() failed
Traceback (most recent call last):
  File "D:\a\cpython\cpython\Lib\test\pythoninfo.py", line 1022, in collect_info
    collect_func(info_add)
    ~~~~~~~~~~~~^^^^^^^^^^
  File "D:\a\cpython\cpython\Lib\test\pythoninfo.py", line [9](https://github.com/python/cpython/actions/runs/7286334537/job/19854864684#step:4:10)34, in collect_windows
    line = output.splitlines()[0]
           ~~~~~~~~~~~~~~~~~~~^^^
IndexError: list index out of range

Collection failed: exit with error
```

https://github.com/python/cpython/actions/runs/7286334537/job/19854864684

More examples:

* https://github.com/python/cpython/actions/runs/7298148666/job/19888624148?pr=113378
* https://github.com/python/cpython/actions/runs/7298041643/job/19888309424?pr=113390
* https://github.com/python/cpython/actions/runs/7297567068/job/19887032026?pr=113389

Sometimes it fails on re-run, sometimes it passes.

Let's output the exit code and error message when it happens, and fail fast, to help debug.

* Issue: gh-113440

### CODE REVIEW

**Comment by hugovk** [comment]
Ah interesting:

I ran this about 25 times on my fork, they passed every time.

First time with this PR it failed :)


```pytb
Running Debug|Win32 interpreter...
ERROR: collect_windows() failed
Traceback (most recent call last):
  File "D:\a\cpython\cpython\Lib\test\pythoninfo.py", line 1025, in collect_info
    collect_func(info_add)
    ~~~~~~~~~~~~^^^^^^^^^^
  File "D:\a\cpython\cpython\Lib\test\pythoninfo.py", line 929, in collect_windows
    raise ValueError(
    ...<2 lines>...
    )
ValueError: Command 'ver' failed with exit code 3221225794: stdout='' stderr=''
```


https://learn.microsoft.com/en-us/azure/service-fabric/service-fabric-diagnostics-code-package-errors says:


Exit code | Hexadecimal value | Short description | Root cause | Potential fix
-- | -- | -- | -- | --
3221225794 | 0xc0000142 | STATUS_DLL_INIT_FAILED | This error sometimes means that the machine has run out of desktop heap space. This cause is especially likely if you have numerous processes that belong to your application running on the node. | If your program wasn't built to respond to Ctrl+C signals, you can enable the **EnableActivateNoWindow** setting in the Cluster manifest. Enabling this setting means your code package will run without a GUI window and won't receive Ctrl+C signals. This action also reduces the amount of desktop heap space each process consumes. If your code package needs to receive Ctrl+C signals, you can increase the size of your node's desktop heap.

File: Lib/test/pythoninfo.py [hunk 1]
```diff
@@ -924,9 +924,12 @@ def collect_windows(info_add):
                                 stdout=subprocess.PIPE,
                                 stderr=subprocess.PIPE,
                                 text=True)
```
Comment (zooba) [inline Lib/test/pythoninfo.py:938]:
```suggestion
                                text=True,
                                env={
                                    "SystemDrive": os.environ["SystemDrive"],
                                    "SystemRoot": os.environ["SystemRoot"],
                                    "Path": r"{0}\system32;{0}".format(os.environ["SystemRoot"]),
                                    "PATHEXT": os.environ["PATHEXT"],
                                })
```

Just an idea to try and avoid issues with GHA's environment? You'll also need an `import os` as there isn't one in this file yet.

**Comment by zooba** [comment]
Looks like it's actually a problem with PATH then. I don't have a problem with leaving that `env` code in there, though it's probably worth a comment.

**Comment by hugovk** [comment]
Would you like to suggest a comment?

**Comment by zooba** [comment]
`Avoid interference from CI systems with messy environments (gh-113392)`?

**Comment by zooba** [comment]
Looks like it's gotten stuck this time, which is very interesting. Perhaps the previous run was on the successful side of the 50/50?

**Comment by hugovk** [comment]
I ran about 20 times on my fork, but it passed each time. 

I'll restart it. But getting stuck is worse than stopping with an error...

**Comment by hugovk** [comment]
I won't restart it as it's already been restarted twice.

* 1st run: stalled and was cancelled https://github.com/python/cpython/actions/runs/7299741157/attempts/1?pr=113392

* 2nd run: pass https://github.com/python/cpython/actions/runs/7299741157/attempts/2?pr=113392

* 3rd run: same 3221225794 error https://github.com/python/cpython/actions/runs/7299741157?pr=113392

```
ERROR: collect_windows() failed
Traceback (most recent call last):
  File "D:\a\cpython\cpython\Lib\test\pythoninfo.py", line 1027, in collect_info
    collect_func(info_add)
    ~~~~~~~~~~~~^^^^^^^^^^
  File "D:\a\cpython\cpython\Lib\test\pythoninfo.py", line 931, in collect_windows
    raise ValueError(
    ...<2 lines>...
    )
ValueError: Command 'ver' failed with exit code 3221225794: stdout='' stderr=''
```

**Comment by zooba** [comment]
Okay, guess we can assume that PATH isn't the culprit (and revert that change).

I've been able to find some (private) discussion about a possible issue that looks like this. It suggests that `STARTUPINFOW.lpDesktop` needs to be initialised to `Winsta0\Default` rather than being left empty for cases where the process is being launched into a new window (which `shell=True` should do) from a service account (which we _shouldn't_ be, but maybe we are?).

Currently it's `_winapi_CreateProcess_impl` in `Modules\_winapi.c` that needs to set that value (in `si.StartupInfo`). It's _probably_ not a real behaviour change to set it unconditionally, but ideally we'd support it on the startup info object. Might be worth making the change just for testing. It might also be possible to repro locally [using psexec](https://stackoverflow.com/a/17973778) to run as NETWORK SERVICE.

A simpler fix might be to pass `CREATE_NO_WINDOW` in the creation flags, which ought to bypass the entire code path that leads to the issue above. But I see no suggestion of that being a fix/workaround (because in the general case where you need a window, it's not).

**Comment by serhiy-storchaka** [comment]
I compared successful and failing logs, and they are identical to this point, ignoring time, commit hash, git pull progress indicator and the order of compiling files.

**Comment by serhiy-storchaka** [comment]
Is there an open issue? I'm trying to ignore this error, if it can help, because it blocks our work. See #113435.

**Comment by hugovk** [comment]
I don't think there's an issue. 

Your PR looks good to unblock, perhaps combine with this to log the error rather than raising an exception?

**Comment by serhiy-storchaka** [comment]
@hugovk Then could you please open an issue and copy your description to it? We will link both PRs to this issue, unblock the workflow, and then continue investigation of the problem.

**Comment by hugovk** [comment]
Certainly, I've opened issue https://github.com/python/cpython/issues/113440.

**Comment by hugovk** [comment]
Do we still need this or can we close it (and https://github.com/python/cpython/issues/113440)?

**Comment by serhiy-storchaka** [comment]
Do you still need this?

The `ver` error no longer blocks CI, but the cause of the failure is still unclear. It would be nice to find and fix the cause, but for now we can just ignore it.

**Pull request closed without merging**